eclass/secureboot.eclass

```diff
@@ -0,0 +1,179 @@
+# argument.
+
+case ${EAPI} in
+	7|8) ;;
+	*) die "${ECLASS}: EAPI ${EAPI:-0} not supported" ;;
+esac
+
+IUSE="secureboot"
+BDEPEND="secureboot? ( app-crypt/sbsigntools )"
+
+# @ECLASS_VARIABLE: SECUREBOOT_SIGN_KEY
+# @USER_VARIABLE
+# @DEFAULT_UNSET
+# @DESCRIPTION:
+# Used with USE=secureboot.  Should be set to the path of the private
+# key in PEM format to use, or a PKCS#11 URI.
+#
+# @ECLASS_VARIABLE: SECUREBOOT_SIGN_CERT
+# @USER_VARIABLE
+# @DEFAULT_UNSET
+# @DESCRIPTION:
+# Used with USE=secureboot.  Should be set to the path of the public
+# key certificate in PEM format to use.
+
+if [[ -z ${_SECUREBOOT_ECLASS} ]]; then
+_SECUREBOOT_ECLASS=1
+
+# @FUNCTION: _secureboot_die_if_unset
+# @INTERNAL
+# @DESCRIPTION:
+# If USE=secureboot is enabled die if the required user variables are unset
+# and die if the keys can't be found.
+_secureboot_die_if_unset() {
+	use secureboot || return
+	debug-print-function ${FUNCNAME[0]} "${@}"
+	if [[ -z ${SECUREBOOT_SIGN_KEY} || -z ${SECUREBOOT_SIGN_CERT} ]]; then
+		die "USE=secureboot enabled but SECUREBOOT_SIGN_KEY and/or SECUREBOOT_SIGN_CERT not set."
+	fi
+	if [[ ! ${SECUREBOOT_SIGN_KEY} == pkcs11:* && ! -f ${SECUREBOOT_SIGN_KEY} ]]; then
+		die "SECUREBOOT_SIGN_KEY=${SECUREBOOT_SIGN_KEY} not found"
+	fi
+	if [[ ! -f ${SECUREBOOT_SIGN_CERT} ]];then
+		die "SECUREBOOT_SIGN_CERT=${SECUREBOOT_SIGN_CERT} not found"
+	fi
+}
+
+# @FUNCTION: secureboot_pkg_setup
+# @DESCRIPTION:
+# Checks if required user variables are set before starting the build
+secureboot_pkg_setup() {
+	use secureboot || return
+	debug-print-function ${FUNCNAME[0]} "${@}"
+	# If we are merging a binary then the files in this binary
+	# are already signed, no need to check the variables.
+	if [[ ${MERGE_TYPE} != binary ]]; then
+		_secureboot_die_if_unset
+	fi
+}
+
+# @FUNCTION: secureboot_sign_efi_file
+# @USAGE: <input file> <output file>
+# @DESCRIPTION:
+# Sign a file using sbsign and the requested key/certificate.
+# If the file is already signed with our key then skip.
+secureboot_sign_efi_file() {
+	use secureboot || return
+	debug-print-function ${FUNCNAME[0]} "${@}"
+	local input_file=${1}
+	local output_file=${2}
+
+	_secureboot_die_if_unset
+
+	if has_version app-crypt/sbsigntools; then
+		ebegin "Signing ${input_file}"
+		local return=1
+		if sbverify "${input_file}" --cert "${SECUREBOOT_SIGN_CERT}" &> /dev/null; then
+			ewarn "${input_file} already signed, skipping"
+			return=0
+		else
+			local args=(
+				"--key=${SECUREBOOT_SIGN_KEY}"
+				"--cert=${SECUREBOOT_SIGN_CERT}"
+			)
+			if [[ ${SECUREBOOT_SIGN_KEY} == pkcs11:* ]]; then
+				args+=( --engine=pkcs11 )
+			fi
+
+			sbsign "${args[@]}" "${input_file}" --output "${output_file}"
+			return=${?}
+		fi
+		eend ${return} || die -n "Signing ${input_file} failed"
+	else
+		die "secureboot signing requested but app-crypt/sbsigntools not installed"
+	fi
+}
+
+# @FUNCTION: secureboot_auto_sign
+# @USAGE: [--in-place | --ext <file extension>]
```

Comment:
FWICS you're not using the `--ext` variant anywhere in this PR. Is there any use case for it right now? If not, let's keep things simple and drop it until it's actually necessary.

Reply:
Not really. I'll just drop it then.